src/ocamlorg_frontend/pages/internal_error.eml

```
@@ -0,0 +1,25 @@
+let render () =
+Layout.render
+~title:"Internal Error"
+~description:"This page does not exist." @@
+<div class="bg-white px-4 py-16 sm:px-6 sm:py-24 md:grid md:place-items-center lg:px-8">
+  <div class="sm:flex max-w-max mx-auto">
+    <p class="text-4xl font-extrabold text-orange-600 sm:text-5xl">500</p>

```

The HTTP status isn't necessarily 500, Dream calls the error handler for errors other than 500. In fact, if I understand correctly, with this change, unfound pages (404) will now use this template?

---

I have been tested and you're corrected 400 will be use this templates. 

I think we need to changes, do you have any suggestions what I need to do?

---

The error template receives `suggested_response ` as a last argument (https://aantron.github.io/dream/#val-error_template). You can use this to get the status (`Dream.status`) and with the status you can also get an explanation of the error using `Dream.status_to_reason`. You can pass these to the templates to display the status and explanation.

---

Currently, the 404 pages are being served from `static.ml`'s `loader` function which returns a 404 response if an asset is not found under the given URL. At the bottom of `router.ml`, we capture `/**` to try and load an asset - after all other routes fail to match.

Looking at the Dream documentation for `error_template`, it says

> In case of a 4xx or 5xx response from the application, that response itself is passed to the template.

Oh, that is interesting. If I understand that correctly, in case of a 4xx response the error handler of Dream will catch that and place the response into `suggested_response`. And indeed, when the handler returns a 4xx or 5xx response, we can be sure that this is using its own dedicated template. In that case, we should just return this suggested response coming from the handler.
It looks like we can see what the cause of the error is by inspecting the `error` parameter: https://aantron.github.io/dream/#type-error.

---

@tmattio passing all errors through the template will not work because this would override the package documentation 404 template.

However, it still makes sense to have the "ordinary" 404 page returned by `http_or_404` use this new generic error template. (This doesn't have to be done in this PR, but could be a later refactor.)

Instead of returning the regular `not_found` response, `http_or_404` could throw an exception that `error_template`'s error handler can consume to render the 404 error into the generic error response. :thinking:

---

> @tmattio passing all errors through the template will not work because this would override the package documentation 404 template.

Make sense. Tbh I don't think that would a problem to have a single error template but if we want to keep a separate template for the documentation what I suggested doesn't work indeed.

Perhaps a more general approach would be to check if suggested_response contains a body already and only use the template if it doesn't.

---

I haven't looked into how exactly Dream behaves wrt `suggested_response`. If we can check for presence of a body that makes sense and would be simpler than the exception approach.

---

To elaborate on why we need the package documentation 404 error page:

1. When you use the version switcher on a documentation page, you currently cannot see whether that page exists in a specific version. When you reach a 404 page this way, it's disorienting.
2. For packages that did not have their documentation build successfully: sometimes we have an error to display for the failed package documentation build, and sometimes we only have a 404 page.
3. Links into the package documentation of the latest version may, on a new release of the package, turn into 404 pages - it's much more useful to offer the user the search bar of the package documentation and the link to the package overview in this situation. A generic 404 page would be disappointing in this context.